### DISCUSSION TITLE
Unable to Connect Unity TCP using Fixed IP from GCP (Google Cloud Platform)

### DISCUSSION THREAD

**Discussion by a-i-am** [community/community#143177] [Programming Help]
### I’m currently working on connecting a Unity TCP client and server using an external IP address obtained from Google Cloud Platform (GCP). However, despite numerous attempts, I am unable to make a successful connection with the fixed IP from GCP. I’m listing below the things I have already checked and the issues that persist. If anyone could point out any overlooked aspects or potential solutions, I would really appreciate it.
----------------------------------------------------
I am experiencing issues with a TCP connection in Unity using a static IP created on Google Cloud Platform (GCP). I attempted to establish a server-client connection using TCP with an external IP assigned by GCP, but after multiple tries, I am still unable to resolve the failure described below, and thus unable to use the external IP. If anyone has suggestions on what I might have overlooked, or if you have potential solutions or insights on the cause, I would greatly appreciate your help.

### Confirmed Functional Components
Internal IP Address Port Forwarding: Verified NAT settings by logging into the router.
Ping Test: Confirmed active network connection by pinging the GCP server’s IP from the client.

**GCP Firewall Rules**:
Source IP Range: 0.0.0.0/0
TCP Protocol: Allowed
Direction: Ingress
Port and Static IP: No issues identified.

**VM Instance Settings**:
Firewall: HTTP and HTTPS traffic allowed.

**GCP SSH Remote Server Access:** Successfully logged into the server with the static IP created by GCP.

**Server Port Check via CMD:** Verified if the port is open on the server using netstat -an | find "{GCP allowed PORT}".
Local Testing with 127.0.0.1: Local testing was successful.

**Issues:**
CMD Telnet {GCP Static IP} {GCP Allowed PORT}:
Result: Unable to connect to host {GCP Static IP}... could not connect to port {GCP Allowed PORT}.
Unity Socket Error:
Result: Unable to connect due to the target computer actively refusing the connection.

Attempts to Disable Windows Firewall:
Tried disabling Windows Firewall but encountered the same failure.
Client Code Execution Delay:
Delayed client code execution by 5 seconds to ensure the server starts first, but no change in outcome.

**Unity Source Code Setup**:
The server and client codes were assigned to separate components on empty GameObjects in Unity.

**Server.cs**

```
using System.Collections.Generic;
using System.Net;
using System.Net.Sockets;
using UnityEngine;
using System;
using System.IO;

public class Server : MonoBehaviour
{
    List<ServerClient> clients;
    List<ServerClient> disconnectList;
    TcpListener server;
    bool serverStarted;
    private const int PORT = ; // Set port number
    private const string IP_ADDRESS = "0.0.0.0"; // Allow connections from all IPs

    public void ServerCreate()
    {
        clients = new List<ServerClient>();
        disconnectList = new List<ServerClient>();

        try
        {
            server = new TcpListener(IPAddress.Parse(IP_ADDRESS), PORT);
            server.Start();
            Debug.Log($"Server started on port {PORT}.");
            StartListening();
            serverStarted = true;
        }
        catch (Exception e)
        {
            Debug.LogException(e);
        }
    }

    void Start()
    {
        ServerCreate();
    }

    void Update()
    {
        if (!serverStarted) return;

        foreach (ServerClient c in clients)
        {
            if (!IsConnected(c.tcp))
            {
                c.tcp.Close();
                disconnectList.Add(c);
                continue;
            }
            else
            {
                NetworkStream s = c.tcp.GetStream();
                if (s.DataAvailable)
                {
                    string data = new StreamReader(s, true).ReadLine();
                    if (data != null)
                        OnIncomingData(c, data);
                }
            }
        }

        for (int i = 0; i < disconnectList.Count; i++)
        {
            Debug.Log($"{disconnectList[i].clientName} disconnected.");
            clients.Remove(disconnectList[i]);
            disconnectList.RemoveAt(i);
        }
    }

    bool IsConnected(TcpClient c)
    {
        try
        {
            if (c != null && c.Client != null && c.Client.Connected)
            {
                if (c.Client.Poll(0, SelectMode.SelectRead))
                    return !(c.Client.Receive(new byte[1], SocketFlags.Peek) == 0);
                return true;
            }
            else
                return false;
        }
        catch
        {
            return false;
        }
    }

    void StartListening()
    {
        server.BeginAcceptTcpClient(AcceptTcpClient, server);
    }

    void AcceptTcpClient(IAsyncResult ar)
    {
        TcpListener listener = (TcpListener)ar.AsyncState;
        ServerClient newClient = new ServerClient(listener.EndAcceptTcpClient(ar));
        clients.Add(newClient);
        Debug.Log($"Client connected: {newClient.clientName}");
        StartListening();
    }

    void OnIncomingData(ServerClient c, string data)
    {
        if (data.Contains("&NAME"))
        {
            c.clientName = data.Split('|')[1];
            Debug.Log($"{c.clientName} connected.");
            return;
        }
    }

    public void Broadcast(string data, List<ServerClient> cl)
    {
        foreach (var c in cl)
        {
            try
            {
                StreamWriter writer = new StreamWriter(c.tcp.GetStream());
                writer.WriteLine(data);
                writer.Flush();
            }
            catch (Exception e)
            {
                Debug.LogException(e);
            }
        }
    }
}

public class ServerClient
{
    public TcpClient tcp;
    public string clientName;

    public ServerClient(TcpClient clientSocket)
    {
        clientName = "Guest";
        tcp = clientSocket;
    }
}
```

**Client.cs**

```
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.Net.Sockets;
using System.IO;
using System;

public class Client : MonoBehaviour
{
    bool socketReady;
    TcpClient socket;
    NetworkStream stream;
    StreamWriter writer;
    StreamReader reader;
    private const string SERVER_IP = " "; // GCP Static IP
    private const int SERVER_PORT = ; // Client port number

    public void ConnectToServer()
    {
        if (socketReady) return;

        try
        {
            socket = new TcpClient(SERVER_IP, SERVER_PORT);
            stream = socket.GetStream();
            writer = new StreamWriter(stream);
            reader = new StreamReader(stream);
            socketReady = true;
            Debug.Log("Connected to server.");
            SendClientName("Guest" + UnityEngine.Random.Range(1000, 10000)); // Random client name
        }
        catch (Exception e)
        {
            Debug.Log($"Socket error: {e.Message}");
        }
    }

    void Start()
    {
        StartCoroutine(ConnectAfterDelay());
    }

    private IEnumerator ConnectAfterDelay()
    {
        yield return new WaitForSeconds(5); // Delay to allow server startup
        ConnectToServer();
    }

    void Update()
    {
        if (socketReady && stream.DataAvailable)
        {
            string data = reader.ReadLine();
            if (data != null)
                Debug.Log(data); // Display received data
        }
    }

    void SendClientName(string clientName)
    {
        Send($"&NAME|{clientName}");
    }

    void Send(string data)
    {
        if (!socketReady) return;

        writer.WriteLine(data);
        writer.Flush();
    }

    void OnApplicationQuit()
    {
        CloseSocket();
    }

    void CloseSocket()
    {
        if (!socketReady) return;

        writer.Close();
        reader.Close();
        socket.Close();
        socketReady = false;
    }
}


```

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by tigran-sargsyan-w** [comment]
I took a close look at your situation, and it sounds like everything internal is working, but the external (“static”) IP + port forwarded from GCP isn’t letting the Unity client in. This usually means something in the network stack or VM config is still blocking TCP connections from outside, even if NAT/router/ping/firewall rules look right. Here are some ideas & a way to test/fix that should help:

From what you describe, these are likely culprits:

- Even though your GCP firewall allows ingress from 0.0.0.0/0 on the TCP port, the VM’s OS‐level firewall (Windows or Linux) may still be blocking that port. Disabling the Windows firewall is good, but make sure the rule is specifically for the port + incoming TCP, not just general “HTTP/HTTPS”.
- The IP address you use in Unity client must be the external static IP of the VM (not local/internal), and the server code must be listening on either IPAddress.Any/0.0.0.0 or the specific interface bound to that external IP. Using "0.0.0.0" is correct for all interfaces, but sometimes VMs have multiple network interfaces and routing/NAT behaves differently.
- Possibly, the router/NAT or GCP network config is not forwarding or mapping correctly — e.g. the external static IP must be assigned to the VM, and then firewall + OS permissions allow the process to accept traffic.

Here’s a test & fix plan that almost always reveals the issue:
1. On the VM itself, run a simple TCP listener (outside of Unity) on that port, e.g. using netcat or something lightweight. From another machine (outside GCP), try to connect to that port (using IP:port). If that fails, the issue is outside Unity, likely OS firewall or GCP network.
2. If the external test works, try using Unity server in “localhost” or binding to the specific external IP on the VM, not just 0.0.0.0, to see if binding matters.
3. Check GCP’s VPC network configuration: ensure there are no extra firewall rules or routing tables that override or restrict ingress for the VM/subnet.
4. If the VM has multiple network interfaces or private + public IPs, check that the public IP maps traffic properly and that the server's response can route back through the same IP/interface (sometimes asymmetric routing or return‐path filtering blocks replies).

Here’s a minimal example of a TCP listener in C# you can use on the server/VM (outside Unity or even within Unity, just for test) to isolate the problem:

```
using System;
using System.Net;
using System.Net.Sockets;
using System.Text;

class SimpleTcpServer
{
    static void Main()
    {
        int port = YOUR_PORT_NUMBER;
        IPAddress ip = IPAddress.Any;  // listens on all interfaces

        TcpListener listener = new TcpListener(ip, port);
        listener.Start();
        Console.WriteLine($"Listening on port {port}...");

        while (true)
        {
            using (TcpClient client = listener.AcceptTcpClient())
            using (NetworkStream stream = client.GetStream())
            {
                Console.WriteLine("Client connected");
                byte[] msg = Encoding.UTF8.GetBytes("Hello from server");
                stream.Write(msg, 0, msg.Length);
            }
        }
    }
}
```
On the client side (outside GCP, on a different network), you can test with telnet {STATIC_IP} {PORT} or a basic C# TCP client to see if it connects.

I believe once you confirm traffic can reach the VM (via the test server), the Unity code will work (because your Unity setup already looks solid: 0.0.0.0 + delay, stream write/read, etc.). If after the external listener works you still get “actively refusing”, then likely the OS firewall is still rejecting or the Unity server process isn’t running or listening at that moment.
